pkg/controller/installation/windows/calico_windows_upgrader.go

```diff
@@ -0,0 +1,438 @@
+// inProgress, or inSync.
+func (w *calicoWindowsUpgrader) getNodeUpgradeStatus() (map[string]*corev1.Node, map[string]*corev1.Node, map[string]*corev1.Node, error) {
+	pending := make(map[string]*corev1.Node)
+	inSync := make(map[string]*corev1.Node)
+	inProgress := make(map[string]*corev1.Node)
+	expectedVersion := w.getExpectedVersion()
+
+	for _, obj := range w.nodeIndexInformer.GetIndexer().List() {
+		node, ok := obj.(*corev1.Node)
+		if !ok {
+			return nil, nil, nil, fmt.Errorf("Never expected index to have anything other than a Node object: %v", obj)
+		}
+
+		if node.Labels[corev1.LabelOSStable] != "windows" {
+			continue
+		}
+
+		windowsLog.V(1).Info(fmt.Sprintf("Processing node %v", node.Name))
+		exists, variant, version := common.GetNodeVariantAndVersion(node)
+
+		// If the version annotation doesn't exist, something is wrong with the
+		// Calico Windows node or it needs to be upgraded manually to
+		// a version supported by the calicoWindowsUpgrader.
+		if !exists {
+			return nil, nil, nil, fmt.Errorf("Node %v does not have the version annotation, it might be unhealthy or it might be running an unsupported Calico version.", node.Name)
+		}
+
+		if node.Labels[common.CalicoWindowsUpgradeLabel] == common.CalicoWindowsUpgradeLabelInProgress {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v has the upgrade in-progress label", node.Name))
+			inProgress[node.Name] = node
+		} else if variant != w.install.Variant || version != expectedVersion {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v doesn't have the latest variant and/or version. variant=%v, expectedVariant=%v, version=%v, expectedVersion=%v", node.Name, variant, w.install.Variant, version, expectedVersion))
+			pending[node.Name] = node
+		} else {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v has the latest variant and version", node.Name))
+			inSync[node.Name] = node
+		}
+	}
+
+	windowsLog.V(1).Info(fmt.Sprintf("pending=%v, in-progress=%v, in-sync=%v", len(pending), len(inProgress), len(inSync)))
+	return pending, inProgress, inSync, nil
+}
+
+func (w *calicoWindowsUpgrader) upgradeCompleted(node *corev1.Node) bool {
+	_, variant, version := common.GetNodeVariantAndVersion(node)
+	if variant == w.install.Variant && version == w.getExpectedVersion() {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) isUpgradeFromCalicoToEnterprise(node *corev1.Node) bool {
+	_, variant, _ := common.GetNodeVariantAndVersion(node)
+	if variant == operatorv1.Calico && w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) getExpectedVersion() string {
+	if w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return components.EnterpriseRelease
+	}
+	return components.CalicoRelease
+}
+
+func (w *calicoWindowsUpgrader) updateWindowsNodes() {
+	pending, inProgress, inSync, err := w.getNodeUpgradeStatus()
+	if err != nil {
```

Comment:
I think this error should be reflected in the status resource. I don't think it is ok to just log this error and continue on.

Reply:
Fixed in 0cc494ff181599152ef9476ba3b661b0f591cfad

Reply:
Also, this error branch should be above - immediately below where the error is generated. Generally bad practice to put additional code between the error handling, since it can lead to unexpected bugs and makes the code harder to read.

Reply:
We discussed offline and have a plan to address the status handling properly in a follow PR.